File: experiments/selenium-java/saucedemo-e2e/src/main/java/com/qaskills/saucedemo/pages/CartPage.java
package com.qaskills.saucedemo.pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class CartPage extends BasePage {

    private static final By PAGE_TITLE = By.cssSelector(".title");
    private static final By CHECKOUT_BUTTON = By.id("checkout");
    private static final By CONTINUE_SHOPPING_BUTTON = By.id("continue-shopping");

    public CartPage(WebDriver driver) {
        super(driver);
    }

    public void waitForPage() {
        waitForUrlContains("cart.html");
        wait.until(driver -> getText(PAGE_TITLE).equals("Your Cart"));
    }

    public boolean isItemInCart(String productName) {
        return isDisplayed(cartItem(productName));
    }

    public CartPage removeItem(String productName) {
        click(By.xpath("//div[contains(@class,'cart_item') and contains(.,'" + productName
                + "')]//button[contains(@data-test,'remove')]"));
        return this;
    }

    public InventoryPage continueShopping() {
        click(CONTINUE_SHOPPING_BUTTON);
        return new InventoryPage(driver);
    }

    public CheckoutPage proceedToCheckout() {
        click(CHECKOUT_BUTTON);
        return new CheckoutPage(driver);
    }

    private By cartItem(String productName) {
        return By.xpath("//div[contains(@class,'cart_item') and contains(.,'" + productName + "')]");
    }
}

File: experiments/cypress-e2e/saucedemo-e2e/cypress/pages/cart.page.ts
export class CartPage {
  get pageTitle() {
    return cy.contains('.title', 'Your Cart');
  }

  get checkoutButton() {
    return cy.getByTest('checkout');
  }

  get continueShoppingButton() {
    return cy.getByTest('continue-shopping');
  }

  assertLoaded() {
    cy.url().should('include', 'cart.html');
    this.pageTitle.should('be.visible');
    return this;
  }

  cartItem(productName: string) {
    return cy.contains('.cart_item', productName);
  }

  assertItemVisible(productName: string) {
    this.cartItem(productName).should('be.visible');
    return this;
  }

  removeItem(productName: string) {
    this.cartItem(productName).contains('button', 'Remove').click();
    return this;
  }

  proceedToCheckout() {
    this.checkoutButton.click();
    return this;
  }
}

export const cartPage = new CartPage();
